src/middlewared/middlewared/plugins/virt/instance.py

```diff
@@ -0,0 +1,451 @@
+        results = (await incus_call('1.0/instances?filter=&recursion=2', 'get'))['metadata']
+        entries = []
+        for i in results:
+            # If entry has no config or state its probably in an unknown state, skip it
+            if not i.get('config') or not i.get('state'):
+                continue
+            entry = {
+                'id': i['name'],
+                'name': i['name'],
+                'type': 'CONTAINER' if i['type'] == 'container' else 'VM',
+                'status': i['state']['status'].upper(),
+                'cpu': i['config'].get('limits.cpu'),
+                'autostart': i['config'].get('boot.autostart') or False,
+                'environment': {},
+                'raw': i,
+                'aliases': [],
+            }
+
+            if memory := i['config'].get('limits.memory'):
+                # Handle all units? e.g. changes done through CLI
+                if memory.endswith('MiB'):
+                    memory = int(memory[:-3]) * 1024 * 1024
+                else:
+                    memory = None
+            entry['memory'] = memory
+
+            for k, v in i['config'].items():
+                if not k.startswith('environment.'):
+                    continue
+                entry['environment'][k[12:]] = v
+            entries.append(entry)
+
+            for v in (i['state']['network'] or {}).values():
+                for address in v['addresses']:
+                    if address['scope'] != 'global':
+                        continue
+                    entry['aliases'].append({
+                        'type': address['family'].upper(),
+                        'address': address['address'],
+                        'netmask': int(address['netmask']),
+                    })
+
+        return filter_list(entries, filters, options)
+
+    @private
+    async def validate(self, new, schema_name, verrors, old=None):
+        if not old and await self.query([('name', '=', new['name'])]):
+            verrors.add(f'{schema_name}.name', f'Name {new["name"]!r} already exists')
+
+        # Do not validate image_choices because its an expansive operation, just fail on creation
+
+    def __data_to_config(self, data):
+        config = {}
+        if data.get('environment'):
+            for k, v in data['environment'].items():
+                config[f'environment.{k}'] = v
+        if data.get('cpu'):
+            config['limits.cpu'] = data['cpu']
+
+        if data.get('memory'):
+            config['limits.memory'] = str(data['memory']) + 'MiB'
+
+        if data.get('autostart') is not None:
+            config['boot.autostart'] = str(data['autostart']).lower()
+        return config
+
+    @api_method(VirtInstanceImageChoicesArgs, VirtInstanceImageChoicesResult)
+    async def image_choices(self, data):
+        """
+        Provice choices for instance image from a remote repository.
+        """
+        choices = {}
+        if data['remote'] in (None, 'LINUX_CONTAINERS'):
```

Comment:
None is the default, in case its not specified.
There plans to add another remote soon.

Reply:
I am confused on that end..if it is `None` or if it is the other value we have - we have the same behavior, so why not use `LINUX_CONTAINERS` as a default?

Reply:
Its a matter a API usage, no? Consumer could send None and it will always use a default we choose, as opposed of having a pre-defined value and consumer will have to change in case of any API change as well instead of using a default.

Tell me your preference and I will make the change.

Reply:
I think it is not very intuitive that we switch remote defaults like this, personally it seems better to me to have it explicit in the schema with a default value of what we want the default to be. We do this in various other places in middleware where we specify our own default to what we think it should be and keep it explicit in schema.
I think we should default to `LINUX_CONTAINERS` and not allow `None` in this case if you agree?